File: environment.cpp
#include "environment.h"
#include <cmath>

void Environment::initialize() {
    for (int lin = 0; lin < HEIGHT; lin++) {
        for (int col = 0; col < WIDTH; col++) {
            if (lin == 0 || lin == HEIGHT-1 || col == 0 || col == WIDTH-1) {
                grid[lin][col] = 2; // Border
            } else {
                grid[lin][col] = 0; // Empty space
            }
        }
    }
    
    const int OBSTACLE_SIZE = 16;
    const int OBSTACLE_POSITIONS[] = {25, 91, 160};
    
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            int lin = OBSTACLE_POSITIONS[i];
            int col = OBSTACLE_POSITIONS[j];
            
            for (int y = lin; y < lin + OBSTACLE_SIZE; y++) {
                for (int x = col; x < col + OBSTACLE_SIZE; x++) {
                    grid[y][x] = 2;
                }
            }
        }
    }
}

bool Environment::isPathClear(double startLin, double startCol, double angle, int steps) const {
    double testlin = startLin - (steps * sin((PI * angle) / 180));
    double testcol = startCol + (steps * cos((PI * angle) / 180));

    if (testlin < 1 || testlin > HEIGHT-2 || testcol < 1 || testcol > WIDTH-2) 
        return false;

    if (grid[(int)testlin][(int)testcol])
        return false;

    return true;
}

void Environment::setCell(int lin, int col, int value) {
    if (lin >= 0 && lin < HEIGHT && col >= 0 && col < WIDTH) {
        grid[lin][col] = value;
    }
}

int Environment::getCell(int lin, int col) const {
    if (lin >= 0 && lin < HEIGHT && col >= 0 && col < WIDTH) {
        return grid[lin][col];
    }
    return -1;
}

File: python_version/environment.py
"""
Environment class for robot navigation.
"""
import numpy as np
import math
from constants import HEIGHT, WIDTH


class Environment:
    def __init__(self):
        """Initialize empty environment grid."""
        self.grid = np.zeros((HEIGHT, WIDTH), dtype=np.int8)
        self.initialize()

    def initialize(self):
        """Initialize environment with borders and obstacles."""
        # Set borders
        self.grid[0, :] = 2
        self.grid[-1, :] = 2
        self.grid[:, 0] = 2
        self.grid[:, -1] = 2

        # Add some obstacles (can be customized)
        # TODO: Add your obstacle pattern here

    def is_path_clear(self, start_lin: float, start_col: float, angle: float, steps: int) -> bool:
        """
        Check if path is clear in the given direction.
        
        Args:
            start_lin: Starting line position
            start_col: Starting column position
            angle: Direction angle in degrees
            steps: Number of steps to check
            
        Returns:
            bool: True if path is clear, False otherwise
        """
        rad_angle = math.radians(angle)
        sin_angle = math.sin(rad_angle)
        cos_angle = math.cos(rad_angle)

        for step in range(steps):
            # Calculate position after step
            lin = int(start_lin + step * sin_angle)
            col = int(start_col + step * cos_angle)

            # Check bounds
            if not (0 <= lin < HEIGHT and 0 <= col < WIDTH):
                return False

            # Check for obstacles
            if self.grid[lin, col] == 2:
                return False

        return True

    def set_cell(self, lin: int, col: int, value: int):
        """Set value of a cell in the grid."""
        if 0 <= lin < HEIGHT and 0 <= col < WIDTH:
            self.grid[lin, col] = value

    def get_cell(self, lin: int, col: int) -> int:
        """Get value of a cell in the grid."""
        if 0 <= lin < HEIGHT and 0 <= col < WIDTH:
            return self.grid[lin, col]
        return 2  # Treat out of bounds as obstacle